lib/streamlit/runtime/pages_manager.py

```diff
@@ -103,16 +104,102 @@ def get_active_script(
     def get_pages(self) -> dict[PageHash, PageInfo]:
         return source_util.get_pages(self.pages_manager.main_script_path)
 
+    def set_pages(self, _pages: dict[PageHash, PageInfo]) -> None:
+        raise NotImplementedError("Unable to set pages in this V1 strategy")
+
+    def get_page_script(self, _fallback_page_hash: PageHash) -> Optional[PageInfo]:
+        raise NotImplementedError("Unable to get page script in this V1 strategy")
+
+
+class PagesStrategyV2:
+    def __init__(self, pages_manager: PagesManager, **kwargs):
+        self.pages_manager = pages_manager
+        self._active_script_hash: PageHash = self.pages_manager.main_script_hash
+        self._pages: dict[PageHash, PageInfo] | None = None
+        self._initial_page_script_hash: PageHash | None = None
+        self._initial_page_name: PageName | None = None
+
+    def get_active_script_hash(self) -> PageHash:
+        return self._active_script_hash
+
+    def set_active_script_hash(self, page_hash: PageHash):
+        self._active_script_hash = page_hash
+
+    def get_initial_active_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> PageInfo:
+        # At this point, we cannot determine the active script at start
+        # as we don't have the list of pages yet. So we will pass
+        # the common code script path and the hash
+        #
+        # We will, however, save the page_script_hash or page name
+        # to be used when the list of pages is received.
+        self._initial_page_script_hash = page_script_hash
+        self._initial_page_name = page_name
+
+        return {
+            # We always run the main script in V2 as it's the common code
+            "script_path": self.pages_manager.main_script_path,
+            "page_script_hash": page_script_hash
+            or self.pages_manager.main_script_hash,  # Default Hash
+        }
```

Comment:
I feel like this should probably be renamed or split into two methods with the V1 version being a no-op. The `get_*` method name that also sets values important for things to work correctly later is quite confusing, especially when it's used as a setter method in test code.